### ISSUE TITLE
Opened Activities Stack is Fundamentally Wrong

### ISSUE THREAD

**Issue by renas** [RobotiumTech/robotium#454]
_From [Marek.Urbaniak](https://code.google.com/u/Marek.Urbaniak/) on January 07, 2013 02:38:25_

My team has a long history creating workarounds in our app because of Robotium leaking the memory. Renas has a long history trying to fix the issues in Robotium. However, the issue is still present in the newest version of Robotium.

Storing references to all opened activities as weak references is not a proper fix. For sure it makes solo.getAllOpenedActivities() unreliable as you may get a list where you will have null values instead of actual activities that have been opened, depending on a memory footprint. Weak references do not solve the out of memory error as when you enter a new activity which intensively allocates a lot of memory, garbage collector will not manage to free weakly reachable objects immediately. It does require a few garbage collection cycles!

If you have tests where you finish activities you would very easily run into OOM exception. It may be because a test is using the back button or that some activities finish by themselves. In our app we have an expensive splash screen (with graphics and animations) which is after 2 seconds finished and ready to be removed from the memory. Robotium keeps this screen in the memory. We had to stop loading and testing any UI of this screen in the test mode because of this issue present in Robotium.

Over the course of 9 months intensive manual testing didn't reproduce OOM error. Although a few testers were trying to do it. Also we don't have any report of OOM error in the Play Store 6 months after the launch. This is not an issue with Android or our app but with Robotium. After analysing memory footprint while running Robotium and without it in Eclipse MAT (and other tools), and after analysing Robotium source code and trying to patch it, we are confident about the cause of our problems. We have more than 300 test cases which exercise our app on the course of 1.5 hour. The issue was never seen in normal app run but when the app is instrumented with Robotium it happens very often, on random basis, making our automated tests irrelevant.

All the functionality you get from solo.getAllOpenedActivities(), you can get by checking opened activity after every test case. This functionality is just not needed and very detrimental. Every time I talk with someone about it they are very surprised and even terrified that Robotium does store the list of activities.

Please remove this detrimental functionality. There's a new UI testing framework released by Google and it would be a waste of time if people had to migrate to it. What steps will reproduce the problem? 1. Testing an app which in a loop opens a heavy activity (for example with an image background) and moves back. What is the expected output? What do you see instead? Test running instead of failing with out of memory exception. What version of the product are you using? On what operating system? Robotium 3.6, Android emulator 2.3.3 but also 4.2

_Original issue: http://code.google.com/p/robotium/issues/detail?id=375_

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on January 07, 2013 05:38:23_

Marek, thank you very much for this. The reason for storing the activities is to finish the ones that are still alive when the test case ends. Otherwise the next test execution will hang. My hope was that weak references would be a good way of enabling finishing of activities that are still alive, while releasing the ones that are not. Obviously it is not good enough. Lets see what we can do to improve this.

**Status:** Accepted  
  **Owner:** renasr...@gmail.com

**Comment by renas** [comment]
_From [Marek.Urbaniak](https://code.google.com/u/Marek.Urbaniak/) on January 07, 2013 07:25:12_

Thanks for accepting the issue, Renas. You can just track the number of activities in the history stack (instead of keeping the entire stack) and then send the appropriate number of the back key events. It's just about letting the system to do what it's supposed to do and finish the activities.

**Comment by renas** [comment]
_From [Marek.Urbaniak](https://code.google.com/u/Marek.Urbaniak/) on January 07, 2013 08:14:52_

Even a simple patch by changing activitySyncTimerTask from:

if(!activity.isFinishing()){
  if(activitiesStoredInActivityStack.remove(activity.toString()))
    removeActivityFromStack(activity);

  addActivityToStack(activity);
}						}

to:

if (activitiesStoredInActivityStack.remove(activity.toString())) {
  removeActivityFromStack(activity);
}
						
if (!activity.isFinishing()){
 addActivityToStack(activity);
}

should do the trick. Then finished activities would not be kept in the memory by Robotium.

However, keeping the stack of not finished activities is fine as currently Android is not shutting down background activities in the stack to regain the memory. Nonetheless, it can be added at some point. System architecture perfectly allows for this.

The bottom line is that this simple fix should work for now perfectly.

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on January 07, 2013 11:39:48_

Great. Thanks again for this.

**Comment by renas** [comment]
_From [Marek.Urbaniak](https://code.google.com/u/Marek.Urbaniak/) on January 09, 2013 09:40:12_

We are using the patched version already and it seems to work OK. However, interestingly, for some tests we had to add mSolo.sleep(500) before calling mSolo.sendKey(KeyEvent.KEYCODE_BACK) because otherwise the event was lost somehow. Really have no clue how this patch could affect it.

**Comment by renas** [comment]
_From [Marek.Urbaniak](https://code.google.com/u/Marek.Urbaniak/) on January 11, 2013 05:54:30_

It has still failed on OOM after applying the fix. However for now it seems that it has brought an improvement -- the percentage of failed builds is lower. Now we have 1 failure after 8 successful tests runs. Maybe this fix is not removing all the finished activities or just removing the entire stack is needed to fix properly the issue.

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on January 16, 2013 00:05:33_

Is that 8 successful test runs in a row before the crash happens? How was it before applying the patch?

**Comment by renas** [comment]
_From [cha...@gmail.com](https://code.google.com/u/118377364589183354775/) on January 18, 2013 06:59:10_

I'm seeing something similar (3.6). It seem most notable with activities which load images. ie like a splash screens. In the app I'm testing I've been avoiding bringing up splash screen for this reason, instead I load my "home" activity directly. However this isn't the same path a user would do, so I am a little hesitant to continue this approach.

(i haven't tried the fix above yet)

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on January 21, 2013 00:05:49_

@chatko please try the fix above and let me know how it goes.

**Comment by renas** [comment]
_From [Marek.Urbaniak](https://code.google.com/u/Marek.Urbaniak/) on January 23, 2013 02:41:41_

As far as I remember before applying the patch it was half-half. However, that part of the build history has been already overwritten with new builds, so I can't check exact numbers. Also there were new tests coming so the probability for running into this heisenbug was increasing.

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on February 26, 2013 22:25:20_

Thanks again for this. It has been included in Robotium 4.0.

**Status:** Verified

**Comment by renas** [comment]
_From [Marek.Urbaniak](https://code.google.com/u/Marek.Urbaniak/) on February 27, 2013 04:18:51_

Thanks Renas. I've just put the new version in place instead of the patched one. In the case our CI reports any new problems, I will let you know.

**Comment by renas** [comment]
_From [harigur...@gmail.com](https://code.google.com/u/102100610624152927072/) on June 25, 2013 19:06:37_

hi renas,

I am getting the same issue after running 40 testcases in my project,i used solo.finish openedactivities() function in tear down.please let me know the solution and it would be greatful.
